### DISCUSSION TITLE
How to get Enum value in OPC

### DISCUSSION THREAD

**Discussion by MatthewBoss96** [FreeOpcUa/opcua-asyncio#1221] [Q&A] [answered]
I am wondering how I can see the value of the enum in OPC UA.
Currently I have:
`async def EnumTest(self, server, idx:int, name:str) -> Struct:
        await new_enum(server, idx, name, [
            "UNDEFINED",
            "GOOD",
            "BAD",
        ])
`

`self.EnumTest(self.server, self.idx, "EnumerateTest")
`

`self.server.load_data_type_definitions() 
 `

`await self.add_variable(self.Test, self.idx, "EnumTest", "Test for enumerate", ua.EnumerateTest.UNDEFINED)`

When I view the EnumTest value in OPC UA I see:
![image](https://user-images.githubusercontent.com/71376345/220284594-a1683d29-8c43-4c58-8336-b697e86a3a51.png)

How do I see the enum as well? So `Value : 0 (UNDEFINED)` instead of `Value : 0`?

**Comment by schroeder-** [comment, marked as answer]
You need to pass the nodeid of your enumtype to add_variable:
``` Python
async def EnumTest(self, server, idx:int, name:str) -> Struct: 
     enum_node = await new_enum(server, idx, name, [ "UNDEFINED", "GOOD", "BAD", ]) 
     self.EnumTest(self.server, self.idx, "EnumerateTest")
     await self.server.load_data_type_definitions()  
     await self.add_variable(self.Test, self.idx, "EnumTest", "Test for enumerate", ua.EnumerateTest.UNDEFINED, enum_node)
```

**Reply by MatthewBoss96** [reply]
Hi @schroeder- ! Thanks for answering, but with the code you provided I get the same result as before where the value is still '0' instead of '0 (UNDEFINED)'?

**Reply by AndreasHeine** [reply]
enums in opcua are just int32! the string values are provided in the DataTypeDefinitions!

![image](https://user-images.githubusercontent.com/56362817/220293284-fd7e8802-43fd-4a57-bdc1-9d9e3392eff6.png)
![image](https://user-images.githubusercontent.com/56362817/220293484-3c252a23-5abd-4cda-85e2-85bbf67b2156.png)

so you can basically scan them and construct your own Enum from that!

not sure if "load_data_type_definitions()" does that for you but i guess @schroeder- would not post it if not ;)

so basically you need to resolve the int value to the enum string value right!?

**Reply by schroeder-** [reply]
@MatthewBoss96 you are missing the await for load_data_type_definitions
``` Python
async def EnumTest(self, server, idx:int, name:str) -> Struct: 
     enum_node = await new_enum(server, idx, name, [ "UNDEFINED", "GOOD", "BAD", ]) 
     self.EnumTest(self.server, self.idx, "EnumerateTest")
     await self.server.load_data_type_definitions()  
     await self.add_variable(self.Test, self.idx, "EnumTest", "Test for enumerate", ua.EnumerateTest.UNDEFINED, enum_node)
```

**Reply by MatthewBoss96** [reply]
@schroeder- I already added the await to the load_data_type_definitions(), but I will try the tips of @AndreasHeine. Thanks both for replying! I will let you know about the progress

**Reply by AndreasHeine** [reply]
@MatthewBoss96 usually you should already have the enum after loading via "load_data_type_definitions()"!

```python
from asyncua import ua

'''
class NamingRuleType(IntEnum):
    """
    https://reference.opcfoundation.org/v104/Core/docs/Part3/8.29
    :ivar Mandatory:
    :vartype Mandatory: 1
    :ivar Optional:
    :vartype Optional: 2
    :ivar Constraint:
    :vartype Constraint: 3
    """
    Mandatory = 1
    Optional = 2
    Constraint = 3
'''

member = ua.NamingRuleType.Mandatory
print("Value:", member.value)
print("String:", member.name)


member = ua.NamingRuleType(1)
print("Value:", member.value)
print("String:", member.name)
```

```
Value: 1
String: Mandatory
Value: 1
String: Mandatory
```

**Reply by MatthewBoss96** [reply]
Goodmorning @schroeder- @AndreasHeine , I just discovered the problem. I was using a custom made function of `add_variable` instead of the in-built `add_variable` of asyncua. The solution was to use the inbuilt version:
```python
enum_node = await new_enum(self.server, self.idx, "EnumerateTest", [ "UNDEFINED", "GOOD", "BAD", ]) 
await self.server.load_data_type_definitions() 
self.enumTest = await self.Test.add_variable(self.idx, "my_enum", ua.EnumerateTest.UNDEFINED, datatype=enum_node.nodeid)
```
Thanks for replying so quickly and being so involved in the community. Keep up the good work!
